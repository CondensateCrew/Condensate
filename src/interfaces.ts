// DATA INTERFACES
export interface UserLoginPosting {
  email: string,
  password: string
}

export interface UserLoginReceived {
  readonly id: number,
  readonly firstName: string,
  readonly lastName: string
}

export interface UserSignupPosting {
  email: string,
  password: string,
  firstName: string,
  lastName: string,
  id?: number
}

export interface Action {
  id: number,
  name: string
}

export interface Brainstorm {
  id: number,
  question: string,
  idea: string,
  action: string,
  isGenius: boolean,
  categories: Category[]
}

export interface Category {
  id: number,
  name: string
}

export interface Insight {
  id: number,
  question: string,
  answers: string[]
}

export interface WordContext {
  type: string,
  connectedWord: string[]
}

export interface RandomWordCollection {
  [key: string]: WordContext
}

export interface AppStore {
  user: UserLoginReceived,
  allBrainstorms: Brainstorm[],
  actions: Action[],
  categories: Category[],
  readonly randomWordCollections: RandomWordCollection[],
  readonly questionTemplates: string[],
  readonly secretSauce: string[],
  currentBrainstorm: Brainstorm,
  chosenWords: string[],
  insights: Insight[],
  query: string,
  filter: string,
  timeEnded: boolean,
  instructionsEnded: boolean
}

// ACTIONS OBJECTS

export interface ActionObject {
  type: string,
  [key: string]: any
}

export interface IAddUserAction {
  type: 'ADD_USER',
  user: UserLoginReceived
}

export interface ILogoutUserAction {
  type: 'LOG_OUT_USER'
}

export interface IAddAllActionsAction {
  type: 'ADD_ALL_ACTIONS',
  actions: Action[]
}

export interface IAddNewActionAction {
  type: 'ADD_NEW_ACTION',
  action: Action
}

export interface IAddAllBrainstormsAction {
  type: 'ADD_ALL_BRAINSTORMS',
  brainstorms: Brainstorm[]
}

export interface IChangeBrainstormAction {
  type: 'TOGGLE_BRAINSTORMS',
  id: number
}

export interface IAddNewBrainstormAction {
  type: 'ADD_BRAINSTORM',
  brainstorm: Brainstorm
}

export interface IDeleteBrainstormAction {
  type: 'DELETE_BRAINSTORM',
  id: number
}

export interface IAddAllCategoriesAction {
  type: 'ADD_ALL_CATEGORIES',
  categories: Category[]
}

export interface IAddNewCategoryAction {
  type: 'ADD_CATEGORY',
  category: Category
}

export interface IAddRandomWordCollectionsAction {
  type: 'ADD_ALL_COLLECTIONS',
  collections: RandomWordCollection[]
}

export interface IAddQuestionTemplatesAction {
  type: 'ADD_ALL_TEMPLATES',
  templates: string[]
}

export interface IAddSecretSauceAction {
  type: 'ADD_SECRETE_SAUSE',
  secretSauce: string[]
}

export interface IAddCurrentBrainstormAction {
  type: 'ADD_CURRENT_BRAINSTORM',
  currentBrainstorm: Brainstorm
}

export interface IRemoveCurrentBrainstormAction {
  type: 'REMOVE_CURRENT_BRAINSTORM'
}

export interface IAddChosenWordAction {
  type: 'ADD_WORD',
  chosenWord: string
}

export interface IRemoveChosenWordAction {
  type: 'REMOVE_WORD',
  chosenWord: string
}

export interface IAddInsightAction {
  type: 'ADD_INSIGHT',
  insight: Insight
}

export interface IAddQueryAction {
  type: 'ADD_QUERY',
  query: string
}

export interface IRemoveQueryAction {
  type: 'REMOVE_QUERY'
}

export interface IAddFilterAction {
  type: 'ADD_FILTER',
  filter: string
}

export interface IRemoveFilterAction {
  type: 'REMOVE_FILTER'
}

export interface IEndTimeAction {
  type: 'END_TIME'
}

export interface IReverseTimeAction {
  type: 'REVERSE_TIME'
}

export interface IEndInstructionsAction {
  type: 'END_INSTRUCTIONS'
}

export interface IReverseInstructionsAction {
  type: 'REVERSE_INSTRUCTIONS'
}
