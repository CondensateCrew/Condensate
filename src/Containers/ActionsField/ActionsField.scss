@import '../../css/variables.scss';
@import '../../css/index.scss';

.action-toggle {
  width: 100%;
  position: relative;
  margin-top: $tiny-space;
  ul {
    @extend %one-line-block;
    padding: $small-space;
    font-size: $tiny-text;
    margin-top: $tiny-space;
    &.clicked {
      position: absolute;
      top: 1.75rem; left: 0;
      width: 100%;
      z-index: 25;
    }
    li {
      cursor: pointer;
      border-bottom: 1px solid $dark-shadow-primary-transparent;
      border-top: 1px solid $dark-shadow-primary-transparent;
      padding: $small-space;
      &:first-child {
        border-top: none;
      }
      &:last-child {
        border-bottom: none;
        padding-bottom: $tiny-space;
      }
      &:hover {
        font-weight: 700;
      }
    }
  }
  header {
    @extend %one-line-block;
    border: 1px solid $light-border;
    font-size: $small-text;
    padding: $small-space;
    cursor: pointer;
    display: -webkit-flex;
    display: -ms-flex;
    display: flex;
    align-items: center;
    justify-content: space-between;
    position: relative;
    z-index: 30;
    img {
      width: $small-text;
    }
  }
}
