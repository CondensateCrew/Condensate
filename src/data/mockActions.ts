const mockActions: any[] = [
  'Create an App',
  'Write a Story',
  'Plan a Lesson',
  'Make a Recipe',
  'Write a Song'
]

export default mockActions;