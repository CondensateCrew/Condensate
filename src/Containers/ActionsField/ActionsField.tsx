import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { AppStore, Action } from 'interfaces';
import './ActionsField.scss';
import down from 'assets/down.svg';
import { IBrainstormForm } from '../../interfaces';
interface Props {
  formState: IBrainstormForm
  setAction: (formState: IBrainstormForm) => void;
}

const ActionsField:React.FC<Props> = ({ formState, setAction }) => {
  const [ selectedAction, setSelectedAction ] = useState<Action>({id: 1, action: 'Create an app'});
  const [ isClicked, setIsClicked ] = useState<boolean>(false);
  const actionsCollection = useSelector((store:AppStore) => store.actions);

  const handleChange = (e:React.MouseEvent<HTMLElement>):void => {
    const currentTarget = e.target as HTMLElement;

    let newAction = actionsCollection.filter((action:Action) => { //eslint-disable-line
        if (action.action === currentTarget.innerText) {
          return action
        }
    });

    setSelectedAction(newAction[0]);
    setIsClicked(false);
  }

  const toggleBlock = (): void => setIsClicked(!isClicked);

  const updateAction = () => {
    let newChosenAction = actionsCollection.filter((action:Action) => { //eslint-disable-line
      if (action.action === selectedAction.action) {
        return action
      }
    })
    setAction({...formState, action: newChosenAction[0]})
  }

  useEffect(updateAction, [ selectedAction ]);

  const actions = actionsCollection.map((action:Action) => {//eslint-disable-line
    if (selectedAction.action.toLowerCase() !== action.action.toLowerCase()) {
      return <li key={action.id} onClick={handleChange}>{action.action}</li>
    }
  });

  const headerState = (isClicked) ? 'clicked' : '';

  return (
    <div className="action-toggle">
      <header className={headerState} onClick={toggleBlock}>
        <p className="chosen">{selectedAction.action}</p>
        <img src={down} alt="down"/>
      </header>
      {isClicked &&
        <ul className={headerState}>
        { actions }
        </ul>
      }
    </div>
  )
}

export default ActionsField;
