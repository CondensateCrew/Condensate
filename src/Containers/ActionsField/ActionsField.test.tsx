/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { IBrainstormForm } from '../../interfaces';
import ActionsField from './ActionsField';
import mockActions from '../../data/mockActions';

describe('ActionsField', () => {
  let wrapper;
  beforeEach(() => {
    interface Props {
      formState: IBrainstormForm,
      setAction: (formState: IBrainstormForm) => void
    }
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    let mockFormState = {
      question: '',
      categories: [],
      action: 'create an app',
      reset: true
    };

    let mockSetAction = jest.fn();

    wrapper = shallow(<ActionsField
      formState={mockFormState}
      setAction={mockSetAction}
    />);

    expect(wrapper).toMatchSnapshot();
  });
  // it('should call setAction when a change event occurs', () => {
  //   let mockFormState = {
  //     question: '',
  //     categories: [],
  //     action: 'create an app',
  //     reset: true
  //   };
  //
  //   let mockSetAction = jest.fn();
  //
  //   wrapper = mount(<ActionsField
  //     formState={mockFormState}
  //     setAction={mockSetAction}
  //   />);
  //
  //   let mockEvent = {target: {value: 'Write a Book'}};
  //
  //   const instance = wrapper.instance();
  //
  //   wrapper.find('select').simulate('change', mockEvent);
  //   expect(mockSetAction).toHaveBeenCalled();
  // });
  // it('should update the local state when change event occurs', () => {
  //   let mockFormState = {
  //     question: '',
  //     categories: [],
  //     action: 'create an app',
  //     reset: true
  //   };
  //
  //   let mockSetAction = jest.fn();
  //
  //   wrapper = mount(<ActionsField
  //     formState={mockFormState}
  //     setAction={mockSetAction}
  //   />);
  //
  //   let mockEvent = {target: {value: 'Write a Book'}};
  //
  //   expect(wrapper.find('select').getDOMNode().value).toEqual('Create an App');
  //   wrapper.find('select').simulate('change', mockEvent);
  //   expect(wrapper).toMatchSnapshot();
  // });
});
