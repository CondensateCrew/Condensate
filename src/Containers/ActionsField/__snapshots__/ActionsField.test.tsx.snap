// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionsField should call setAction/handleChange when a new li is clicked 1`] = `
<ActionsField
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setAction={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "action": Object {
              "action": "Write a Song",
              "id": 5,
            },
            "categories": Array [],
            "question": "",
            "reset": false,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="action-toggle"
  >
    <header
      className=""
      onClick={[Function]}
    >
      <p
        className="chosen"
      >
        Write a Song
      </p>
      <img
        alt="down"
        src="down.svg"
      />
    </header>
  </div>
</ActionsField>
`;

exports[`ActionsField should display more actions when down arrow is clicked 1`] = `
<ActionsField
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setAction={[MockFunction]}
>
  <div
    className="action-toggle"
  >
    <header
      className="clicked"
      onClick={[Function]}
    >
      <p
        className="chosen"
      >
        Create an app
      </p>
      <img
        alt="down"
        src="down.svg"
      />
    </header>
    <ul
      className="clicked"
    >
      <li
        key="2"
        onClick={[Function]}
      >
        Write a Story
      </li>
      <li
        key="3"
        onClick={[Function]}
      >
        Plan a Lesson
      </li>
      <li
        key="4"
        onClick={[Function]}
      >
        Make a Recipe
      </li>
      <li
        key="5"
        onClick={[Function]}
      >
        Write a Song
      </li>
    </ul>
  </div>
</ActionsField>
`;

exports[`ActionsField should match the snapshot 1`] = `
<ActionsField
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setAction={[MockFunction]}
>
  <div
    className="action-toggle"
  >
    <header
      className=""
      onClick={[Function]}
    >
      <p
        className="chosen"
      >
        Create an app
      </p>
      <img
        alt="down"
        src="down.svg"
      />
    </header>
  </div>
</ActionsField>
`;

exports[`ActionsField should reset the local state when formState.reset is true 1`] = `
<ActionsField
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setAction={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "action": Object {
              "action": "Write a Story",
              "id": 2,
            },
            "categories": Array [],
            "question": "",
            "reset": false,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="action-toggle"
  >
    <header
      className=""
      onClick={[Function]}
    >
      <p
        className="chosen"
      >
        Write a Story
      </p>
      <img
        alt="down"
        src="down.svg"
      />
    </header>
  </div>
</ActionsField>
`;

exports[`ActionsField should reset the local state when formState.reset is true 2`] = `
<ActionsField
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": true,
    }
  }
  setAction={[MockFunction]}
>
  <div
    className="action-toggle"
  >
    <header
      className=""
      onClick={[Function]}
    >
      <p
        className="chosen"
      >
        Create an app
      </p>
      <img
        alt="down"
        src="down.svg"
      />
    </header>
  </div>
</ActionsField>
`;
